securesystemslib/storage.py

```diff
@@ -0,0 +1,219 @@
+
+
+
+class StorageBackendInterface():
+  """
+  <Purpose>
+  Defines an interface for abstract storage operations which can be implemented
+  for a variety of storage solutions, such as remote and local filesystems.
+  """
+
+  __metaclass__ = abc.ABCMeta
+
+
+  @abc.abstractmethod
+  def get(self, filepath):
+    """
+    <Purpose>
+      Retrieve a file from the storage backend.
+
+    <Arguments>
+      filepath:
+        The full path of the file to be retrieved.
+
+    <Exceptions>
+      securesystemslib.exceptions.StorageError, if the file does not exist or is
+      no accessible.
+
+    <Returns>
+      A file-like object of the file at 'filepath'.
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+  @abc.abstractmethod
+  def put(self, fileobj, filepath):
+    """
+    <Purpose>
+      Store a file-like object in the storage backend.
+
+    <Arguments>
+      fileobj:
+        The file-like object to be stored.
+
+      filepath:
+        The full path to the location where 'fileobj' will be stored.
+
+    <Exceptions>
+      securesystemslib.exceptions.StorageError, if the file can not be stored.
+
+    <Returns>
+      None
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+  @abc.abstractmethod
+  def getsize(self, filepath):
+    """
+    <Purpose>
+      Retrieve the size, in bytes, of the file at 'filepath'.
+
+    <Arguments>
+      filepath:
+        The full path to the file.
+
+    <Exceptions>
+      securesystemslib.exceptions.StorageError, if the file does not exist or is
+      not accessible.
+
+    <Returns>
+      The size in bytes of the file at 'filepath'.
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+  @abc.abstractmethod
+  def create_folder(self, filepath):
+    """
+    <Purpose>
+      Create a folder at filepath and ensure all intermediate components of the
+      path exist.
+
+    <Arguments>
+      filepath:
+        The full path of the folder to be created.
+
+    <Exceptions>
+      securesystemslib.exceptions.StorageError, if the folder can not be
+      created.
+
+    <Returns>
+      None
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+  @abc.abstractmethod
+  def list_folder(self, filepath):
+    """
+    <Purpose>
+      List the contents of the folder at 'filepath'.
+
+    <Arguments>
+      filepath:
+        The full path of the folder to be listed.
+
+    <Exceptions>
+      securesystemslib.exceptions.StorageError, if the file does not exist or is
+      not accessible.
+
+    <Returns>
+      A list containing the names of the files in the folder. May be an empty
+      list.
+    """
+    raise NotImplementedError # pragma: no cover
+
+
+
+
+
+class FilesystemBackend(StorageBackendInterface):
+  """
+  <Purpose>
+    A concrete implementation of StorageBackendInterface which interacts with
+    local filesystems using Python standard library functions.
+  """
+
+  def get(self, filepath):
```

Comment:
Proposal: Can we turn this into a context manager? Right now, this requires the following API pattern:

```python
foo = backend.get("foo")
# operations
foo.close()
```

...which makes it easy for a consumer to accidentally leak a file descriptor. With [contextlib](https://docs.python.org/3/library/contextlib.html), we could do something like this:

```python
with backend.get("foo") as foo:
  # operations...
```

Reply:
Great suggestion, thanks. I want to spend a bit of time with the documentation to figure out a clean way to do this without _requiring_ use of a context manager. I would like to be able to support the current explicit `get()`/`close()` pattern as well as the `with get() as ...` pattern.

Would you expect the interface to indicate/recommend implementing a context manager? When I looked during initial work on this PR I couldn't see an obvious way to do that beyond documentation.

Reply:
I've converted `get()` to a ContextManager in both the interface (docstrings) and the `FilesystemBackend`. As implemented `get()` can only be used as a context manager. I'd appreciate any thoughts on that.